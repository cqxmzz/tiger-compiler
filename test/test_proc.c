/*
 * Test the syscalls associated with processes.
 *
 * Xiangru Chen
 */

#include "stdio.h"
#include "stdlib.h"
#include "coffgrader.h"

int main() {

  char* args[10];
  int pid[20];
  int children = 10;

  args[0] = "";
  args[2] = 0;

  char tmp[20];

  int i;
  for (i = 0; i < children; ++i) {
    sprintf(tmp, "%d", i);
    args[1] = tmp;
    pid[i] = exec("test_proc_child.coff", 2, args);
    assertTrue(pid[i] > 0);
  }

  printf("I'm the father.\n");

  V(0);

  int status;
  for (i = children - 1; i >= 0; --i) {
    assertTrue(join(pid[i], &status) == 1);
    assertTrue(status == 100 + i);
  }
  assertTrue(join(0, &status) == -1);

  int mustfail = exec("mustfail.coff", 0, 0);
  assertTrue(join(mustfail, &status) == 0);

  assertTrue(exec("file_not_exist", 0, 0) == -1);

  printf("seems ok.\n");

  done();

  return 0;
}
