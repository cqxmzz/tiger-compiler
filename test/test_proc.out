I'm the father.
I'm child 0.
I'm child 1.
I'm child 2.
I'm child 3.
I'm child 4.
I'm child 5.
I'm child 6.
I'm child 7.
I'm child 8.
I'm child 9.
seems ok.
